### PR TITLE
Limit GQ to fix GVCF block merging, throw exception for GQB > MAX_GENOTYPE_QUAL

### PR DESCRIPTION
**Pull request by ronlevine**
Implements the gatk part of https://github.com/broadinstitute/gatk-protected/issues/602.

### CODE REVIEW

**Comment by ronlevine** [comment]
@samuelklee Please review.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 1]
```diff
@@ -1439,4 +1439,32 @@ public static void setFilteredGenotypeToNocall(final VariantContextBuilder build
         // update genotypes
         builder.genotypes(genotypes);
     }
+
+    /**
+     * Calculate the Genotype Quality (GQ) by subtracting the second smallest Phred-scaled likelihoods (PL) first smallest.
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Thanks for adding javadoc!  But "subtracting the smallest Phred-scaled likelihood (PL) from the second smallest" is more accurate.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 2]
```diff
@@ -1439,4 +1439,32 @@ public static void setFilteredGenotypeToNocall(final VariantContextBuilder build
...
+
+        int first = plValues[0];
+        int second = plValues[1];
+        if (first > second) {
+            second = first;
+            first = plValues[1];
+        }
+        for (int i = 2; i < plValues.length; i++) {
+            final int candidate = plValues[i];
+            if (candidate >= second) continue;
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: I think we are typically more strict about adherence to the Google Java Style guide in gatk-protected, so I would add braces to all the blocks here (even if optional).

File: src/main/java/org/broadinstitute/hellbender/utils/variant/writers/GVCFWriter.java [hunk 3]
```diff
@@ -71,15 +73,20 @@ public GVCFWriter(final VariantContextWriter underlyingWriter, final List<Intege
...
     static RangeMap<Integer,Range<Integer>> parsePartitions(final List<Integer> gqPartitions) {
         Utils.nonEmpty(gqPartitions);
-        Utils.containsNoNull(gqPartitions, "gqPartitions contains a null Integer");
+        Utils.containsNoNull(gqPartitions, "GQ partitions contains a null Integer");
         final RangeMap<Integer, Range<Integer>> result = TreeRangeMap.create();
         int lastThreshold = 0;
         for (final Integer value : gqPartitions) {
-            if (value < lastThreshold) {
-                throw new IllegalArgumentException("gqPartitions is out of order.  Last is " + lastThreshold + " but next is " + value);
+            if (value < 0) {
```
Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/writers/GVCFWriter.java]: I edited the exception messages in https://github.com/broadinstitute/gsa-unstable/pull/1440.  Mind porting those over?  The old messages were somewhat confusing and not always grammatically correct...

Comment (samuelklee) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/writers/GVCFWriter.java]: See also comments by @SHuang-Broad in his review of that PR.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 1 above)
Comment (ronlevine) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Done

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 2 above)
Comment (ronlevine) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java]: Done

File: src/main/java/org/broadinstitute/hellbender/utils/variant/writers/GVCFWriter.java (hunk 3 above)
Comment (ronlevine) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/writers/GVCFWriter.java]: Done. The first comment is too inefficient due to extra looping. The second comment is not applicable to the current code.

**Comment by ronlevine** [comment]
@samuelklee Comments addressed. Back to you.

**Pull request merged**